tests/zkevm/test_worst_bytecode.py

```diff
@@ -204,3 +204,100 @@ def test_worst_bytecode_single_opcode(
         ],
         exclude_full_post_state_in_output=True,
     )
+
+
+@pytest.mark.valid_from("Cancun")
+@pytest.mark.parametrize("initcode_size", [MAX_INITCODE_SIZE])
+@pytest.mark.parametrize(
+    "pattern",
+    [
+        Op.STOP,
+        Op.JUMPDEST,
+        Op.PUSH1[bytes(Op.JUMPDEST)],
+        Op.PUSH2[bytes(Op.JUMPDEST + Op.JUMPDEST)],
+        Op.PUSH1[bytes(Op.JUMPDEST)] + Op.JUMPDEST,
+        Op.PUSH2[bytes(Op.JUMPDEST + Op.JUMPDEST)] + Op.JUMPDEST,
+    ],
+    ids=lambda x: x.hex(),
+)
+def test_worst_initcode_jumpdest_analysis(
+    state_test: StateTestFiller,
+    fork: Fork,
+    pre: Alloc,
+    initcode_size: int,
+    pattern: Bytecode,
+):
+    """
+    Test the jumpdest analysis performance of the initcode.
+
+    This benchmark places a very long initcode in the memory and then invoke CREATE instructions
+    with this initcode up to the block gas limit. The initcode itself has minimal execution time
+    but forces the EVM to perform the full jumpdest analysis on the parametrized byte pattern.
+    The initicode is modified by mixing-in the returned create address between CREATE invocations
+    to prevent caching.
+    """
+    # Expand the initcode pattern to the transaction data so it can be used in CALLDATACOPY
+    # in the main contract. TODO: tune the tx_data_len param.
+    tx_data_len = 1024
+    tx_data = pattern * (tx_data_len // len(pattern))
+    tx_data += (tx_data_len - len(tx_data)) * bytes(Op.JUMPDEST)
+    assert len(tx_data) == tx_data_len
+    assert initcode_size % len(tx_data) == 0
+
+    # Prepare the initcode in memory.
+    code_prepare_initcode = sum(
+        (
+            Op.CALLDATACOPY(dest_offset=i * len(tx_data), offset=0, size=Op.CALLDATASIZE)
+            for i in range(initcode_size // len(tx_data))
+        ),
+        Bytecode(),
+    )
+
+    # At the start of the initcode execution, jump to the last opcode.
+    # This forces EVM to do the full jumpdest analysis.
+    initcode_prefix = Op.JUMP(initcode_size - 1)
+    code_prepare_initcode += Op.MSTORE(
+        0, Op.PUSH32[bytes(initcode_prefix).ljust(32, bytes(Op.JUMPDEST))]
+    )
+
+    # Make sure the last opcode in the initcode is JUMPDEST.
+    code_prepare_initcode += Op.MSTORE(initcode_size - 32, Op.PUSH32[bytes(Op.JUMPDEST) * 32])
+
+    # We are using the Paris fork where there is no initcode size limit.
+    # This allows us to test the increased initcode limits.
+    # However, Paris doesn't have PUSH0 so simulate it with PUSH1.
+    push0 = Op.PUSH0 if Op.PUSH0 in fork.valid_opcodes() else Op.PUSH1[0]
+
+    code_invoke_create = (
+        Op.PUSH1[len(initcode_prefix)]
+        + Op.MSTORE
```

Comment:
Clever use of the returned address!

Reply:
Yes this was so well disguised that I initially thought it did not edit the initcode :smile: :+1: Love this!